### PR TITLE
Stack release

### PR DESCRIPTION
**Pull request by ProPheT777**
fighter

This version **unlock** StackPHP.

**Key improvments**

* Container and config are insulated
* Container does not require request anymore
* Config is immutable after the start of container
* You can use resolved parameters from container definition directly (no need to write DICT-IT extension)
* Debugger is working at compile time (no need to look at logs application)
* Handle & recover from php fatal error

**WHY**

Currently we need to encapsulate application for many reason, based on stack layer it's not reliable to use event dispatcher (onRequest, onResponse) across layer whereas we use stack philosophy. 

Example : 

```php
<?php
require_once '../vendor/autoload.php';

use Pyrite\Errors\ErrorSubscriber;
use Pyrite\Errors\ErrorSubscription;
use Pyrite\Utils\Logger;
use Stack\Builder;
use Pyrite\Kernel\PyriteKernel as PyriteKernel;
use Symfony\Component\Debug\Debug;
use Symfony\Component\HttpFoundation\Request;
use Pyrite\Routing\Router;
use Pyrite\Routing\RouteConfigurationBuilderI18n;
use Symfony\Component\HttpFoundation\Response;

date_default_timezone_set('Europe/Paris');

$routingConfig = __DIR__.'/../compiled/config/routes.php';
$containerConfig = __DIR__.'/../compiled/config/container.php';

$router = new Router($routeBuilder = new RouteConfigurationBuilderI18n($routingConfig));
$pyrite = new PyriteKernel($containerConfig, $router, new ErrorSubscriber([
    'Pyrite\PyRest\Exception\ForbiddenException' => new ErrorSubscription('http-error', Response::HTTP_FORBIDDEN),
    'Symfony\Component\Routing\Exception\ResourceNotFoundException' => new ErrorSubscription('http-error', Response::HTTP_NOT_FOUND),
    'Symfony\Component\HttpKernel\Exception\NotFoundHttpException' => new ErrorSubscription('http-error', Response::HTTP_NOT_FOUND),
    'Pyrite\PyRest\Exception\MethodNotAllowedException' => new ErrorSubscription('http-error', Response::HTTP_METHOD_NOT_ALLOWED),
    'Symfony\Component\HttpKernel\Exception\MethodNotAllowedHttpException' => new ErrorSubscription('http-error', Response::HTTP_METHOD_NOT_ALLOWED),
]));

$loggerFactory = $pyrite->getLoggerFactory();
$config = $pyrite->getConfig();
$slack = $config->get('slack');

if(true === $config->get('debug')){
    // Choose your favorite debugger
    Debug::enable();
    //$pyrite->push('Pyrite\KernelStack\ErrorMiddleware', true, true);
}

Logger::attachSentry($loggerFactory, $config->get('sentry_callback'), $config->get('debug'));
Logger::attachSlack($loggerFactory, $slack['token'], $slack['channel'], $slack['username'], $config->get('debug'));

$request = Request::createFromGlobals();

$pyrite->push('Evaneos\Middleware\MultiSiteMiddleware', $config, $loggerFactory);
$pyrite->push('Evaneos\Middleware\I18nMiddleware', $config, $routeBuilder, $loggerFactory);
$pyrite->push('Pyrite\KernelStack\SessionMiddleware', $config);
$pyrite->push('Pyrite\KernelStack\HttpMiddleware', $router, $routeBuilder, $pyrite);
$pyrite->push('Pyrite\KernelStack\LoggerMiddleware', $loggerFactory, $router->getRouteCollection());
$pyrite->push('Pyrite\KernelStack\PyriteMiddleware', $pyrite, $router); // here start the container
$pyrite->push('Pyrite\KernelStack\NewrelicMiddleware', 'www-pyrite');

$pyrite->run($request);



```

### CODE REVIEW

**Comment by oliviermadre** [comment]
Globally awesome, I'll put some comment here and there in the code anyway

File: src/KernelStack/I18nMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+     * @param Request $request
+     * @param int     $type
+     * @param bool    $catch
+     *
+     * @return Response
+     */
+    public function handle(Request $request, $type = self::MASTER_REQUEST, $catch = true)
+    {
+        $host  = $request->getHttpHost();
+        $requestedLocal = str_replace('.', '_', $host);
```
Comment (oliviermadre) [inline src/KernelStack/I18nMiddleware.php]: This is evaneos-centric, that middleware should be in the Evaneos repo

File: src/KernelStack/MultiSiteMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * @param Request $request
+     * @param int     $type
+     * @param bool    $catch
+     */
+    public function handle(Request $request, $type = self::MASTER_REQUEST, $catch = true)
+    {
+        $host = $request->getHttpHost();
+
+        $sites = $this->config->get('default_site_id');
```
Comment (oliviermadre) [inline src/KernelStack/MultiSiteMiddleware.php]: This is evaneos-centric, that middleware should be in the Evaneos repo

File: src/KernelStack/SessionMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+     * @param bool    $catch
+     *
+     * @return Response
+     * @throws \Exception
+     */
+    public function handle(Request $request, $type = self::MASTER_REQUEST, $catch = true)
+    {
+        $parameters = $this->config->get('cookie_parameters');
+
+        if(!isset($parameters[$this->config->get('current_locale')])){
```
Comment (oliviermadre) [inline src/KernelStack/SessionMiddleware.php]: beware that the current_locale is evaneos centric, so is "cookie_parameters".
We should inject those as parameters for the session construction (that's why we have Factories in front of our stacks atm)

File: src/KernelStack/HttpMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        $container = $this->kernel->startContainer();
+
+        $director = new Director($request, $this->routingConfigPath);
+
+        $config = $this->kernel->getConfig();
+
+        $currentLocal = $config->get('current_locale');
+        $availableLocal = $config->get('available_locales');
+
+        if(null !== $currentLocal && null !== $availableLocal){
```
Comment (oliviermadre) [inline src/KernelStack/HttpMiddleware.php]: Choosing the RouteConfigurationBuilder shouldn't be handled by HttpMiddleware

currentLocal(e) & availableLocal(e) are Evaneos-centric, you shouldn't rely on them in the HttpMiddleware (supposed to be generic)

**Comment by ProPheT777** [comment]
* MultisiteMiddleware and I18nmiddleware are out of pyrite scope.
* RouteDirector has been removed

A new example of bootstrap : 

```php
<?php
require_once '../vendor/autoload.php';

use Stack\Builder;
use Pyrite\Kernel\PyriteKernel as PyriteKernel;
use Symfony\Component\Debug\Debug;
use Symfony\Component\HttpFoundation\Request;
use Pyrite\Routing\Router;
use Pyrite\Routing\RouteConfigurationBuilderI18n;

date_default_timezone_set('Europe/Paris');

$routingConfig = '../compiled/config/routes.php';
$containerConfig = '../compiled/config/container.php';

$router = new Router($routeBuilder = new RouteConfigurationBuilderI18n($routingConfig));
$pyrite = new PyriteKernel($containerConfig, $router);

$config = $pyrite->getConfig();
$builder = new Builder();

if(true === $config->get('debug')){
    // Choose your favorite debugger
    Debug::enable();
    //$builder->push('Pyrite\KernelStack\ErrorMiddleware', true, true);
}

$request = Request::createFromGlobals();

$builder->push('Evaneos\Middleware\MultiSiteMiddleware', $config, $pyrite->getLoggerFactory());
$builder->push('Evaneos\Middleware\I18nMiddleware', $config, $routeBuilder, $pyrite->getLoggerFactory());
$builder->push('Pyrite\KernelStack\SessionMiddleware', $config);
$builder->push('Pyrite\KernelStack\HttpMiddleware', $router, $routeBuilder);
$builder->push('Pyrite\KernelStack\LoggerMiddleware', $pyrite->getLoggerFactory(), $router->getRouteCollection());
$builder->push('Pyrite\KernelStack\PyriteMiddleware', $pyrite, $router); // here start the container
$builder->push('Pyrite\KernelStack\NewrelicMiddleware', 'www-pyrite');

$app = $builder->resolve($pyrite);

\Stack\run($app, $request);
```

**Pull request merged**